synapse/types/handlers/sliding_sync.py

```diff
@@ -159,14 +511,20 @@ def deep_copy(self) -> "RoomSyncConfig":
 
     def combine_room_sync_config(
         self, other_room_sync_config: "RoomSyncConfig"
-    ) -> None:
+    ) -> "RoomSyncConfig":
         """
-        Combine this `RoomSyncConfig` with another `RoomSyncConfig` and take the
+        Combine this `RoomSyncConfig` with another `RoomSyncConfig` and return the
         superset union of the two.
         """
+        timeline_limit = self.timeline_limit
+        required_state_map = {
+            event_type: set(state_keys)
+            for event_type, state_keys in self.required_state_map.items()
+        }
+
         # Take the highest timeline limit
         if self.timeline_limit < other_room_sync_config.timeline_limit:
```

Comment:
To get rid of the `self` usage in the body after we assign everything at the top:

```suggestion
        if timeline_limit < other_room_sync_config.timeline_limit:
```